local-tests/chainrunner/chain.py

```diff
@@ -0,0 +1,108 @@
+import os
+import os.path as op
+import subprocess
+
+from .node import node
+from .utils import flags_from_dict
+
+
+# Type alias for int
+class Seq(int): pass
+
+
+class Chain:
+    """Chain is a class for orchestrating setting up and interaction with a local aleph-node blockchain.
+    The constructor takes only one argument - a path to a directory with the workspace.
+    All other parameters are adjusted with `bootstrap()` and `set_flags()`."""
+    def __init__(self, workdir):
+        if not op.exists(workdir):
+            os.makedirs(workdir)
+        if not op.isdir(workdir):
+            raise OSError("given path is not a directory")
+        self.path = op.abspath(workdir)
+
+
+    def __getitem__(self, i):
+        return self.nodes[i]
+
+
+    def __iter__(self):
+        return iter(self.nodes)
+
+
+    def bootstrap(self, binary, accounts, **kwargs):
+        """Bootstrap the chain. `accounts` should be a list of strings.
+        Flags `--account-ids`, `--base-path` and `--raw` are added automatically.
+        All other flags are taken from kwargs"""
+        if not op.isfile(binary):
+            raise FileNotFoundError(f'file not found: {binary}')
+        cmd = [binary, 'bootstrap-chain', '--base-path', self.path, '--account-ids', ','.join(accounts), '--raw']
+        cmd += flags_from_dict(kwargs)
+
+        chainspec = op.join(self.path, 'chainspec.json')
+        with open(chainspec, 'w') as f:
+            subprocess.run(cmd, stdout=f).check_returncode()
+
+        self.nodes = []
+        for a in accounts:
+            n = node(binary, chainspec, op.join(self.path, a), self.path)
+            n.flags['node-key-file'] = op.join(self.path, a, 'p2p_secret')
+            self.nodes.append(n)
+
+
+    def set_flags(self, *args, **kwargs):
+        """Set common flags for all nodes. Positional arguments are used as binary flags and should be strings.
+        Keyword arguments are translated to valued flags: `my_arg=some_val` results in `--my-arg some_val`
+        in the binary call. Seq (type alias for int) can be used to specify numerical values that should
+        be different for each node. `val=Seq(13)` results in `--val 13` for node0, `--val 14` for node1 and so on."""
+        for k in args:
+            for n in self.nodes:
+                n.flags[k] = True
+        for k,v in kwargs.items():
+            for i,n in enumerate(self.nodes):
+                n.flags[k] = v+i if isinstance(v, Seq) else v
+
+
+    def set_binary(self, binary, nodes=None):
+        """Replace nodes' binary with `binary`. Optional `nodes` argument can be used to specify
+        which nodes are affected and should be a list of integer indices (0..N-1). Affects all nodes if ommitted."""
+        if not op.isfile(binary):
+            raise FileNotFoundError(f'file not found: {binary}')
+        idx = nodes or range(len(self.nodes))
+        for i in idx:
+            self.nodes[i].binary = binary
```

Comment:
Nope, `__get_item__` is meant as a convenience shortcut for the "outside" user. I don't want to rely on it internally in case the semantics of it changes in the future.

Reply:
anyway, for me `chain[idx]` means block at height `idx`, not a node `idx`, so I would remove it

up to you, feel free to ignore